### ISSUE TITLE
PSR2 multi-line detection doesn't work for inline IF and string concats

### ISSUE THREAD

**Issue by VasekPurchart** [squizlabs/PHP_CodeSniffer#412]
After upgrading to 2.x, I started to see these two use cases, which were not reported by phpcs before:
``` php
$this->setFoo(true
	? 1
	: 2);
```
``` php
$this->setFoo('some'
	. 'long'
	. 'text');
```

These are both considered as "multi-line function calls" and an `Indent` as well as `CloseBracketLine` errors are reported. The `Indent` error is not reported in 1.5.x, so this might be an BC break as well.

I believe that none of these two situations should actually be considered as "multi-line function calls" in the context of PSR2, it states: 

*Argument lists MAY be split across multiple lines, where each subsequent line is indented once. When doing so, the first item in the list MUST be on the next line, and there MUST be only one argument per line.* https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md#46-method-and-function-calls

In both these examples there is only one argument given for the called method and the indentation which is introduced is present only to provide better readability (the real world cases are with much longer lines and more complex code, of course). So the indentation which is being checked should be that one of the internal structure of the argument (probably by another sniff), not the structure of the function call itself. The same applies to checking of the closing bracket, because this is actually not an multi-line call.

What do you think about this?

**Comment by orlangur** [comment]
I think error reported is perfectly valid. Already saw similar discussion before: http://pear.php.net/bugs/bug.php?id=19768. As to me, PSR-2 statement is not vague at all.

> because this is actually not an multi-line call

I see function call placed across multiple lines. How it can be "not a multi-line call"?

Argument list MAY be split. Argument list MAY be not split. If it is split, indentation MUST correspond to requirements. Argument list MAY consist of 1, 2, 3, ... items.

**Comment by aik099** [comment]
To me the multi-line function call is when each parameter must be placed on a separate line. For such edge cases (as in your examples) where single parameter value is split across multiple lines I really don't know what to do.

Let's wait and see what @gsherwood has to say about that.

**Comment by orlangur** [comment]
> To me the multi-line function call is when each parameter must be placed on a separate line

In PSR-2 there is no rule regarding "... parameter MUST be placed ...". Parameters MAY be placed on multiple lines and if so, there MUST be one parameter per line with proper indentation. So, "multi-line function call" not to be confused with "function call with more than one parameter".

Ok, I don't see how to explain it clearer anyway. @gsherwood answered in older ticket I mentioned, so, resolution is quite predictable :wink: 

A bit more interesting case is when the ultimate parameter is an array: https://github.com/php-fig/fig-standards/issues/61#issuecomment-11973252 Didn't check if it is reported as an error by phpcs currently, as to me it can be treated as an error according to PSR-2 interpreted literally, but I would like to have an ability to use such more compact variant.

**Comment by orlangur** [comment]
Hmmm, seems like things have been changed a bit after the discussion I mentioned. Long thread: https://github.com/zendframework/zf2/issues/4390#issuecomment-17370338

Explanation added after voting by FIG: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide-meta.md
So, line breaks inside arrays and closures are allowed.

Examples shown by topic starter are not explicitly justified but probably it makes sense to ignore any line breaks inside an argument by parity of reasoning.

**Comment by gsherwood** [comment]
The multi-line argument section was ambiguous, so PHP-FIG added some errata to explain it: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide-meta.md#31---multi-line-arguments-09082013

I changed PHPCS to ignore multi-line arrays, function calls and closures, but didn't account for multi-line string concatenation. I don't think PHP-FIG considered that either as it is typically something done outside the function. But still, it is a single argument and should be ignored, so this is a bug in PHPCS.

**Comment by gsherwood** [comment]
Your inline IF statement is a little more complex, but I'll see what I can do there. Again, I'm sure PHP-FIG never considered that people would write code like that, but the errata seems to cover it anyway.

**Comment by VasekPurchart** [comment]
@gsherwood thanks for pointing out the errata, I haven't noticed that before.

If you want more test data or want to test a solution please let me know, we have quite a large code base where it is used.

**Comment by VasekPurchart** [comment]
Thanks for the changes, but I am afraid there might be some more work needed on this, when I tested this on my codebase today I got a stuck progress, now I have spent quite some time narrowing down where it gets stuck and it starts right on 961057c.

It does probably not handle well recursed call in some situations, the simpliest not working use-case would be this:

``` php
<?php

foo(bar(), $a);
```

When I remove the `$a` or the second call it works. Should I create a separate issue for this?

**Comment by gsherwood** [comment]
@VasekPurchart Thank you very much for testing this change. I'll look into that code now. No need to submit another issue for it.

**Comment by gsherwood** [comment]
I've just committed a fix for this. I've run it over PHPCS itself and am no longer getting any infinite loops. Thanks again for testing and reporting it.
